File: pat-static/src/main/java/com/rectanglel/patstatic/patterns/PatternDataManager.java
package com.rectanglel.patstatic.patterns;

import com.google.gson.reflect.TypeToken;
import com.rectanglel.patstatic.model.AbstractDataManager;
import com.rectanglel.patstatic.model.RetrofitPatApi;
import com.rectanglel.patstatic.model.StaticData;
import com.rectanglel.patstatic.patterns.response.PatternResponse;
import com.rectanglel.patstatic.patterns.response.Ptr;

import java.io.File;
import java.io.IOException;
import java.util.List;

import rx.Observable;
import rx.exceptions.Exceptions;

/**
 * /**
 * <p>Create a data manager for patternSelections that will handle:</p>
 * <ul>
 *     <li>getting data from disk</li>
 *     <li>or... getting from retrofit and saving that data to disk</li>
 *     <li>cache clear logic (not done yet)</li>
 * </ul>
 * <p>Created by epicstar on 9/18/16.</p>
 * @author Jeremy Jao
 * @since 78
 */
public class PatternDataManager extends AbstractDataManager<List<Ptr>> {

    private RetrofitPatApi patApiClient;

    public PatternDataManager(File dataDirectory, RetrofitPatApi patApiClient, StaticData staticData) {
        super(dataDirectory, staticData, new TypeToken<List<Ptr>>() {}.getType());
        this.patApiClient = patApiClient;
    }

    @Override
    protected String getCacheFolderName() {
        return "lineinfo";
    }

    private File getPatternsFile(String rt) {
        return new File(getDataDirectory(),
                String.format("%s.json", rt));
    }

    public Observable<List<Ptr>> getPatterns(String rt) {
        File polylineFile = getPatternsFile(rt);
        if (polylineFile.exists()) {
            return getPatternsFromDisk(rt);
        } else {
            return getPatternsFromInternet(rt);
        }
    }

    Observable<List<Ptr>> getPatternsFromDisk(String rt) {
        return Observable.just(getPatternsFile(rt))
                .map(file -> {
                    try {
                        return getFromDisk(file);
                    } catch (IOException e) {
                        throw Exceptions.propagate(e);
                    }
                });
    }

    Observable<List<Ptr>> getPatternsFromInternet(String rt) {
        return patApiClient.getPatterns(rt)
                .map(PatternResponse::getPatternResponse)
                .map(bustimePatternResponse -> {
                    try {
                        List<Ptr> patterns = bustimePatternResponse.getPtr();
                        File patternsFile = getPatternsFile(rt);
                        saveAsJson(patterns, patternsFile);
                        return patterns;
                    } catch (IOException e) {
                        throw Exceptions.propagate(e);
                    }
                }).onErrorResumeNext(throwable -> getPatternsFromDisk(rt));
    }
}

File: pat-static/src/main/kotlin/com/rectanglel/patstatic/patterns/PatternDataManager.kt
package com.rectanglel.patstatic.patterns

import com.google.gson.reflect.TypeToken
import com.rectanglel.patstatic.model.AbstractDataManager
import com.rectanglel.patstatic.model.RetrofitPatApi
import com.rectanglel.patstatic.model.StaticData
import com.rectanglel.patstatic.patterns.response.PatternResponse
import com.rectanglel.patstatic.patterns.response.Ptr
import com.rectanglel.patstatic.routes.BusRoute
import com.rectanglel.patstatic.wrappers.WifiChecker
import io.reactivex.Flowable
import io.reactivex.exceptions.Exceptions
import java.io.File
import java.io.IOException

private val dataType = object : TypeToken<List<Ptr>>() {}.type!!
private const val maxFileAge = 24 * 60 * 1000 * 1000 // 24 hours

/**
 * Create a data manager for pattern selections that will handle:
 *
 * 1. getting data from:
 *     1. disk
 *     2. [RetrofitPatApi] and saving that data to disk
 * 2. cache clear logic
 *
 * Created by epicstar on 8/19/17.
 * @author Jeremy Jao
 */
class PatternDataManager(dataDirectory: File,
                         private val patApiClient: RetrofitPatApi,
                         staticData: StaticData,
                         private val wifiChecker: WifiChecker)
    : AbstractDataManager<List<Ptr>>(
        dataDirectory,
        staticData,
        dataType,
        "lineinfo") {

    fun getPatterns(route: BusRoute): Flowable<List<Ptr>> {
        val polylineFile = getPatternsFile(route.number)
        // 1. if doesn't exist, get from internet (merely a fallback to when a pattern isn't yet saved to disk)
        // 2. if off wifi, always get from disk
        // 3. if age is past 24 hours and there is wifi, retrieve from the internet
        // 4. otherwise,
        if (!polylineFile.exists()) {
            return getPatternsFromInternet(route)
        }
        if (!wifiChecker.isWifiOn()) {
            return getPatternsFromDisk(route.number)
        }
        val polylineAge = polylineFile.lastModified()
        val getNowDate = System.currentTimeMillis()
        return if (getNowDate - polylineAge >= maxFileAge) {
            getPatternsFromInternet(route)
        } else {
            getPatternsFromDisk(route.number)
                .concatMap { ptrList ->
                    when (ptrList?.size) {
                        0 -> getPatternsFromInternet(route)
                        else -> Flowable.just(ptrList)
                    }
                }
        }
    }

    private fun getPatternsFile(rt: String): File = File(dataDirectory, "$rt.json")

    internal fun getPatternsFromDisk(rt: String): Flowable<List<Ptr>> {
        return Flowable.just(getPatternsFile(rt))
                .map { file ->
                    try {
                        getFromDisk(file)
                    } catch (e: IOException) {
                        throw Exceptions.propagate(e)
                    }
                }
    }

    // TODO: remove this warning suppression when ViewModel has fixed
    @Suppress("RedundantVisibilityModifier")
    public fun getPatternsFromInternet(route: BusRoute): Flowable<List<Ptr>> {
        println("whats updog -> $route")
        return patApiClient.getPatterns(route.number, route.datafeed)
                .map(PatternResponse::patternResponse)
                .map { bustimePatternResponse ->
                    try {
                        println("oh hello!")
                        val patterns = bustimePatternResponse.ptr // TODO: FIX
                        val patternsFile = getPatternsFile(route.number)
                        saveAsJson(patterns, patternsFile)
                        patterns
                    } catch (e: IOException) {
                        println("OH NOOO")
                        throw Exceptions.propagate(e)
                    }
                }
                .onErrorResumeNext { _: Throwable -> getPatternsFromDisk(route.number) }
//                .onErrorResumeNext(t -> return getPatternsFromDisk(rt))
    }


}